imap_processing/codice/codice_l2.py

```diff
@@ -866,6 +907,125 @@ def process_hi_sectored(dependencies: ProcessingInputCollection) -> xr.Dataset:
+    # axis
+    ssd_id = l2_dataset["ssd_id"].values
+    valid_ssd = (ssd_id <= 15) & (ssd_id >= 0)
+
+    elevation = nan_array.copy()
+    elevation[valid_ssd] = SSD_ID_TO_ELEVATION[ssd_id[valid_ssd]]
+    l2_dataset["elevation_angle"] = xr.DataArray(
+        data=elevation.astype(np.float32), dims=l2_dataset["ssd_id"].dims
+    )
+    # Calculate ssd energy in meV
+    gain = l2_dataset["gain"].values
+    ssd_energy = l2_dataset["ssd_energy"].values
+    valid_mask = (
+        (np.isin(gain, list(GAIN_ID_TO_STR.keys())))
+        & valid_ssd
+        & (ssd_energy != len(energy_table))
+    )
+    # The columns are organized in order of id and gains
+    # E.g. ssd 0 - LG, ssd 0 - MG, ssd 0 - HG, ssd 1 - LG, ssd 1 - MG, ssd 1 - HG, ...
+    cols = ssd_id * 3 + (gain - 1)
+    ssd_energy_converted = nan_array.copy()
+    ssd_energy_converted[valid_mask] = energy_table[
+        ssd_energy[valid_mask], cols[valid_mask]
+    ]
+    l2_dataset["ssd_energy"].data = ssd_energy_converted.astype(np.float32)
+
+    # Convert spin_sector to spin_angle in degrees
+    theta_angles = nan_array.copy()
+    theta_angles[valid_ssd] = SSD_ID_TO_SPIN_ANGLE[ssd_id[valid_ssd]]
+    l2_dataset["spin_angle"] = (
+        (theta_angles + 15.0 * l2_dataset["spin_sector"]) % 360.0
+    ).astype(np.float32)
+
+    # Calculate TOF in ns
+    tof = l2_dataset["tof"].values
+    # Get valid TOF indices for lookup
+    valid_tof_mask = tof < tof_table.shape[0]
+    tof_ns = nan_array.copy()
+    # Get tof values in ns from first column of tof_table
+    tof_ns[valid_tof_mask] = tof_table[tof[valid_tof_mask], 0]
+    l2_dataset["tof"] = xr.DataArray(
+        data=tof_ns,
+        dims=l2_dataset["tof"].dims,
+    ).astype(np.float32)
+
+    # Calculate energy per nuc
+    energy_nuc = nan_array.copy()
+    # Get value from second column of tof_table (E/n (MeV/n))
+    energy_nuc[valid_tof_mask] = tof_table[tof[valid_tof_mask], 1]
+    l2_dataset["energy_per_nuc"] = xr.DataArray(
+        data=energy_nuc,
+        dims=l2_dataset["tof"].dims,
+    ).astype(np.float32)
+    # Drop unused variables
+    vars_to_drop = ["spare", "sw_bias_gain_mode", "st_bias_gain_mode"]
+    l2_dataset = l2_dataset.drop_vars(vars_to_drop)
+    # Update variable attributes
+    l2_dataset.attrs.update(
+        cdf_attrs.get_global_attributes("imap_codice_l2_hi-direct-events")
+    )
+    for var in l2_dataset.data_vars:
+        l2_dataset[var].attrs.update(cdf_attrs.get_variable_attributes(var))
+    # Update coord attributes
+    l2_dataset["priority"].attrs.update(
+        cdf_attrs.get_variable_attributes("priority", check_schema=False)
+    )
+    l2_dataset["event_num"].attrs.update(
+        cdf_attrs.get_variable_attributes("event_num", check_schema=False)
+    )
+    l2_dataset["epoch"] = xr.DataArray(
+        l2_dataset["epoch"].data,
+        dims="epoch",
+        attrs=cdf_attrs.get_variable_attributes("epoch", check_schema=False),
+    )
```

Comment:
I'm not sure why epoch is getting re-instantiated using the same data. Otherwise can you follow the above pattern...
```suggestion
    for coord in l2_dataset.coords:
        l2_dataset[coord].attrs.update(cdf_attrs.get_variable_attributes(coord), check_schema=False)
```